main_test.go

```diff
@@ -2,59 +2,241 @@ package main
+	}
+	s.pubSubSettings = settings
 }
 
-func (b *FirehoseBackendMock) AckMessage(ctx context.Context, providerMetadata interface{}) error {
-	return nil
+func (s *GcpTestSuite) AfterTest(suiteName, testName string) {
+	s.server.Stop()
 }
 
-func (b *FirehoseBackendMock) RequeueFirehoseDLQ(ctx context.Context, numMessages uint32, visibilityTimeout time.Duration) error {
-	return nil
-}
-func (b *FirehoseBackendMock) UploadFile(ctx context.Context, data []byte, uploadBucket string, uploadLocation string) error {
-	return nil
+func (s *GcpTestSuite) TestNewFirehose() {
+	gcpSettings := gcp.Settings{}
+	var hedwigLogger hedwig.Logger
+	backend := gcp.NewBackend(gcpSettings, hedwigLogger)
+	msgList := []hedwig.MessageTypeMajorVersion{{
+		MessageType:  "user-created",
+		MajorVersion: 1,
+	}}
+	var s3 ProcessSettings
+	var s2 gcp.Settings
+	storageBackend := firehoseGcp.NewBackend(&storage.Client{})
+	encoderDecoder := firehoseProtobuf.FirehoseEncoderDecoder{}
+	f, err := NewFirehose(backend, &encoderDecoder, msgList, storageBackend, s2, s3, hedwigLogger)
+	assert.Equal(s.T(), nil, err)
+	assert.NotNil(s.T(), f)
 }
 
-func (b *FirehoseBackendMock) ReadFile(ctx context.Context, readBucket string, readLocation string) ([]byte, error) {
-	return []byte("abc"), nil
+func (s *GcpTestSuite) TestFirehoseFollowerIntegration() {
+	var hedwigLogger hedwig.Logger
+	backend := gcp.NewBackend(s.pubSubSettings, hedwigLogger)
+	// maybe just user-created?
+	msgList := []hedwig.MessageTypeMajorVersion{{
+		MessageType:  "user-created",
+		MajorVersion: 1,
+	}}
+	s3 := ProcessSettings{
+		MetadataBucket: "some-metadata-bucket",
+		StagingBucket:  "some-staging-bucket",
+		OutputBucket:   "some-output-bucket",
+		FlushAfter:     2,
+	}
+	var s2 gcp.Settings
+	storageBackend := firehoseGcp.NewBackend(s.storageClient)
+	msgTypeUrls := map[hedwig.MessageTypeMajorVersion]string{
+		{MessageType: "user-created", MajorVersion: 1}: "type.googleapis.com/standardbase.hedwig.UserCreatedV1",
+	}
+	encoderDecoder := firehoseProtobuf.NewFirehoseEncodeDecoder(msgTypeUrls)
+	f, err := NewFirehose(backend, encoderDecoder, msgList, storageBackend, s2, s3, hedwigLogger)
+	s.Require().NoError(err)
+
+	routing := map[hedwig.MessageTypeMajorVersion]string{
+		{
+			MessageType:  "user-created",
+			MajorVersion: 1,
+		}: "dev-user-created-v1",
+	}
+	pubEncoderDecoder, err := protobuf.NewMessageEncoderDecoder(
+		[]proto.Message{&UserCreatedV1{}},
+	)
+	s.Require().NoError(err)
+	publisher := hedwig.NewPublisher(backend, pubEncoderDecoder, routing)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+
+	defer cancel()
+
+	userId := "C_1234567890123456"
+	data := UserCreatedV1{UserId: &userId}
+	message, err := hedwig.NewMessage("user-created", "1.0", map[string]string{"foo": "bar"}, &data, "myapp")
+	s.Require().NoError(err)
+	_, err = publisher.Publish(ctx, message)
+	s.Require().NoError(err)
+
+	_, err = publisher.Publish(ctx, message)
```

Comment:
slight modification to message before publishing would be nice for later verification.